### PR TITLE
Fix AutoSlash middleware when a service is mounted prefixed

### LINKED ISSUE #1378: AutoSlash not working when prefix is specified in mountService
I tested this in http4s 0.18.0-M1

How to reproduce:
```
package com.example.http4squickstart

import io.circe._
import org.http4s._
import org.http4s.circe._
import org.http4s.server._
import org.http4s.dsl._
import cats.effect.IO

object HelloWorld {
  val service = HttpService[IO] {
    case GET -> Root / "hello" / name =>
      Ok(Json.obj("message" -> Json.fromString(s"Hello, ${name}")))
  }
}
```

```
package com.example.http4squickstart

import cats.effect.IO
import fs2.Stream

import scala.util.Properties.envOrNone
import org.http4s.util.StreamApp
import org.http4s.server.blaze.BlazeBuilder
import org.http4s.server.middleware.AutoSlash

object BlazeExample extends StreamApp[IO] {

  val port : Int              = envOrNone("HTTP_PORT") map (_.toInt) getOrElse 8080
  val ip   : String           = "0.0.0.0"

  override def stream(args: List[String], requestShutdown: IO[Unit]): Stream[IO, Nothing] =
    BlazeBuilder[IO]
      .bindHttp(port, ip)
      .mountService(AutoSlash(HelloWorld.service), "/test/")
      .serve

}
```

`curl -v http://localhost:8080/test/hello/alessandro` works.
```
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET /test/hello/alessandro HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Date: Fri, 08 Sep 2017 17:01:52 GMT
< Content-Length: 31
< 
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
{"message":"Hello, alessandro"}
```

`curl -v http://localhost:8080/test/hello/alessandro/` doesn't.
```
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET /test/hello/alessandro/ HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Date: Fri, 08 Sep 2017 17:02:33 GMT
< Content-Length: 0
< 
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
```

If I remove the prefix and curl without the prefix AutoSlash works.

### PR DESCRIPTION
**Pull request by aeons**
Fixes #1378

### CODE REVIEW

File: server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala [hunk 1]
```diff
@@ -16,12 +15,13 @@ object AutoSlash {
   def apply[F[_]](service: HttpService[F])(implicit F: Monad[F]): HttpService[F] = Service.lift {
     req =>
       service(req).flatMap {
-        case _: Pass[F] =>
+        case Pass() =>
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala]: This means the same thing but avoids a check against an erased type, right?

File: server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala [hunk 2]
```diff
@@ -16,12 +15,13 @@ object AutoSlash {
...
           val p = req.uri.path
           if (p.isEmpty || p.charAt(p.length - 1) != '/')
-            F.pure(Pass[F])
+            Pass.pure[F]
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala]: This is stylistic, right?

File: server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala [hunk 3]
```diff
@@ -16,12 +15,13 @@ object AutoSlash {
...
           else {
-            val withSlash = req.withUri(req.uri.copy(path = p.substring(0, p.length - 1)))
+            val caret = req.attributes(Request.Keys.PathInfoCaret)
+            val withSlash = req.withUri(req.uri.copy(path = p.substring(caret, p.length - 1)))
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala]: Doesn't this lose the scriptName part of the path?  Should this be `req.withPathInfo`?

I'll bet this is also broken on older versions.

File: server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala (hunk 1 above)
Comment (aeons) [reply server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala]: I think so yes, it's also to be consistent with other places in the code.

File: server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala (hunk 2 above)
Comment (aeons) [reply server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala]: Yes, and for consistency.

File: server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala (hunk 3 above)
Comment (aeons) [reply server/src/main/scala/org/http4s/server/middleware/AutoSlash.scala]: Maybe, I don't know how `pathInfo` works.

The problem right now is that the prefixed part is back in the URI when it's passed to a service that doesn't expect it.

I guess it should `get` and `getOrElse` on the attribute as well.

**Pull request merged**